lib/smack/NormalizeLoops.cpp

```diff
@@ -0,0 +1,134 @@
+#include "llvm/IR/Dominators.h"
+#include <map>
+#include <set>
+
+namespace smack {
+
+using namespace llvm;
+
+// Register LoopInfo
+void NormalizeLoops::getAnalysisUsage(AnalysisUsage& AU) const {
+  AU.addRequired<LoopInfoWrapperPass>();
+}
+
+/**
+ * Creates a basic block in \param func which unconditionally branches to
+ * \param target.
+ */
+BasicBlock* makeForwardingBlock(BasicBlock* target, Function* func) {
+  auto& context = target->getContext();
+  BasicBlock* result = BasicBlock::Create(context, "forwarder", func);
+  BranchInst* branch = BranchInst::Create(target);
+  result->getInstList().push_back(branch);
+  return result;
+}
+
+/**
+ * Splits any edge between \param BB and \param headerBlock in a loop body,
+ * by adding a forwarding block between \param BB and \param headerBlock. The
+ * \param blockMap is updated to map between \param BB and the newly created
+ * forwarding block.
+ */
+void splitBlock(BasicBlock* BB, BasicBlock* headerBlock, Function* func,
+                std::map<BasicBlock*, BasicBlock*>& blockMap) {
+  TerminatorInst* ti = BB->getTerminator();
+  unsigned numSuccs = ti->getNumSuccessors();
+  if (numSuccs <= 1) { return; }
+  BasicBlock* forwarder = nullptr;
+  for (unsigned i = 0; i < numSuccs; ++i) {
+    BasicBlock* succ = ti->getSuccessor(i);
+    if (succ != headerBlock) { continue; }
+
+    // Successor is the header block and this block has multiple successors,
+    // so it needs a forwarding block.
+    if (forwarder == nullptr) {
+      forwarder = makeForwardingBlock(headerBlock, func);
+      blockMap[BB] = forwarder;
+    }
+    ti->setSuccessor(i, forwarder);
+  }
+}
+
+/**
+ * \param blockMap a map from old BasicBlocks and new BasicBlocks.
+ * Fixes any PHI nodes in \param phis to get their incoming block from the new
+ * blocks in \param blockMap.
+ */
+void processPhis(std::vector<PHINode*>& phis, std::map<BasicBlock*, BasicBlock*>& blockMap) {
+  for (PHINode* phi : phis) {
+    unsigned numIncoming = phi->getNumIncomingValues();
+    for (unsigned i = 0; i < numIncoming; ++i) {
+      BasicBlock* incomingBlock = phi->getIncomingBlock(i);
+      if (blockMap.count(incomingBlock)) {
+        phi->setIncomingBlock(i, blockMap[incomingBlock]);
+      }
+    }
+  }
+}
+
+void processLoop(Loop *loop, Function* func) {
+  std::vector<PHINode*> phis;
+  std::set<BasicBlock*> phiIncBlocks;
+  BasicBlock* headerBlock = loop->getHeader();
+
+  // Gather Phis in the header
+  for (Instruction& I : *headerBlock) {
+    if (auto* phi = dyn_cast<PHINode>(&I)) {
+      phis.push_back(phi);
+      phiIncBlocks.insert(phi->block_begin(), phi->block_end());
+    }
+  }
+
+  std::map<BasicBlock*, BasicBlock*> blockMap;
+
+  // Add forwarding blocks
+  for (BasicBlock* BB : phiIncBlocks) {
+    splitBlock(BB, headerBlock, func, blockMap);
+  }
+
+  // Fix up Phi nodes
+  processPhis(phis, blockMap);
+
+  // Handle subloops
+  for (Loop *subloop : loop->getSubLoops()) {
+    processLoop(subloop, func);
```

Comment:
Just curious: have we tested nested loops, or even thought about the possibility of what happens when the processing the nested loop splits blocks? For instance, whether this code ends up processing the original (unspilt) blocks of the enclosing loop, or the updated (split) blocks?

Reply:
I've tested it with nested loops, both handwritten and compiled, and loop headers appear to be unique to the loop. (Su- loops have different headers.) However, the documentation doesn't say that this behavior is guaranteed.